source/common/http/codec_client.cc

```diff
@@ -96,7 +96,8 @@ void CodecClient::onReset(ActiveRequest& request, StreamResetReason reason) {
   deleteRequest(request);
 }
 
-void CodecClient::onData(Buffer::Instance& data) {
+void CodecClient::onData(Buffer::Instance& data, bool end_stream) {
+  remote_closed_ |= end_stream;
```

Comment:
This is definitely the correct fix, but AFAIK we don't use half close semantics for any HTTP code, so I don't think this will actually solve your problem in practice. @ggreenway?

Reply:
Yeah, Network::ReadFilter::onData() should never pass end_stream==true on the http code. If you're seeing that, there's a bug somewhere, probably in ConnectionImpl.

Reply:
Will do some test with real env that reproduced #2715, though I believe this at least make this [if statement](https://github.com/envoyproxy/envoy/blob/c932e50c34a11dd382dca9c5daf2026d4686daa9/source/common/http/http1/conn_pool.cc#L204) true, so it won't assign next pending request in HTTP1 conn pool.

Reply:
https://github.com/envoyproxy/envoy/blob/master/source/common/network/connection_impl.cc#L435

We could consider I guess setting end_stream to true in this case, even for not half-closed connections. That might allow you to win the race in more cases, but I think that is a more involved change.

Reply:
Can ConnectionImpl::onRead() result in a new request being sent out on the upstream connection (if we read the end of a response)? If not, then it shouldn't matter; the RemoteClose is raised right after onRead().

Reply:
@ggreenway that's exactly why we're seeing 503s in #2715 (the code is [here](https://github.com/envoyproxy/envoy/blob/c932e50c34a11dd382dca9c5daf2026d4686daa9/source/common/http/http1/conn_pool.cc#L212)). Let me take more look on how I can fix this.

Reply:
Did this exact case ever work better? Given that onData and onEvent were always separate calls, there was a choice of either calling onEvent(RemoteClose) first, which would mean an error on the response (because we didn't process all the response data yet), or calling onData() first and getting this error.

Reply:
For HTTP/1.1, I don't think this ever worked any better, because upstream is supposed to set `Connection: close` as I keep saying into the void. For HTTP/1 this used to work OK because of https://github.com/envoyproxy/envoy/pull/2743/files#diff-181fe1d44460531b055c9b316ca2cd69R64.

I don't have any objection per say to attempting to handle this better by potentially setting end_stream = true in this case, but again as I've said a ton of times, this is not going to work out in your favor 100% of the time due to certain local/remote interleavings.

Reply:
Sorry for the late response (was sick last week).

I ended up by making onData set end_stream = true even half-close is not enabled. This new parameter is not used widely yet so it should be safe. Letting a TCP read filter know if the byte is the last one (before connection close) is useful even it doesn't handle half-close semantics.

Reply:
whether it's a regression fix or just a fix is somewhat irrelevant no ? envoy should handle "unexpectedly" closed connection without attributing the error to the next request?